jdbc/src/main/scala/org/apache/spark/sql/sources/jdbcExtensions.scala

```diff
@@ -167,26 +192,86 @@ object JdbcExtendedUtils extends Logging {
   }
 
   def getTableWithSchema(table: String, conn: Connection,
-      session: Option[() => SnappySession] = None): (String, String) = {
+      session: Option[() => SparkSession] = None): (String, String) = {
     val dotIndex = table.indexOf('.')
     val schemaName = if (dotIndex > 0) {
       table.substring(0, dotIndex)
-    } else {
+    } else session match {
+      case None => conn.getSchema
       // get the current schema
-      if (session.isDefined) session.get().getCurrentSchema else conn.getSchema
+      case Some(s) => s().catalog.currentDatabase
     }
     val tableName = if (dotIndex > 0) table.substring(dotIndex + 1) else table
     (schemaName, tableName)
   }
 
-  private def getTableMetadataResultSet (table: String, conn: Connection): ResultSet = {
+  private def getTableMetadataResultSet(table: String, conn: Connection): ResultSet = {
     val (schemaName, tableName) = getTableWithSchema(table, conn)
-      // using the JDBC meta-data API
+    // using the JDBC meta-data API
     conn.getMetaData.getTables(null, schemaName, tableName, null)
   }
 
-  def tableExistsInMetaData(table: String, conn: Connection,
-      dialect: JdbcDialect): Boolean = {
+  protected lazy val getCatalystTypeMethod: Method = {
+    val m = JdbcUtils.getClass.getDeclaredMethod(
+      "org$apache$spark$sql$execution$datasources$jdbc$JdbcUtils$$getCatalystType",
+      classOf[Int], classOf[Int], classOf[Int], classOf[Boolean])
+    m.setAccessible(true)
+    m
+  }
+
+  /**
+   * Is the given JDBC type (java.sql.Types) data type signed.
+   */
+  def isSigned(typeId: Int): Boolean = {
+    typeId == Types.INTEGER || typeId == Types.FLOAT || typeId == Types.DECIMAL ||
+        typeId == Types.SMALLINT || typeId == Types.BIGINT || typeId == Types.TINYINT ||
+        typeId == Types.NUMERIC || typeId == Types.REAL || typeId == Types.DOUBLE
+  }
+
+  def getTableSchema(schemaName: String, tableName: String, conn: Connection): Seq[StructField] = {
```

Comment:
Yeah, this seemed to be the right thing to do to find out metadata from the standard jdbc apis. What was the reason we did not do this to start with.

Reply:
This is required only for the case of pre-resolution of table schema to avoid double execution. Otherwise JDBCRDD.resolveTable uses the schema query (which was the problem) and resolves to StructType using the ResultSetMetadata. Perhaps can remove this and instead use "select * from table limit 1" query instead of getColumns.